test/metabase/lib/column_group_test.cljc

```diff
@@ -413,3 +413,43 @@
                {:lib/desired-column-alias "PEOPLE__via__USER_ID__LATITUDE_2", :fk-join-alias "Mock orders card"}
                {:lib/desired-column-alias "PEOPLE__via__USER_ID__CREATED_AT_2", :fk-join-alias "Mock orders card"}]
               (::lib.column-group/columns product-2))))))
+
+(deftest ^:parallel column-group-order-test
+  (testing "column groups should be returned in order: main, explicit joins, implicit joins"
+    (let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+                    (lib/join (lib/join-clause (meta/table-metadata :orders)
+                                               [(lib/= (meta/field-metadata :orders :id)
+                                                       (-> (meta/field-metadata :orders :id)
+                                                           (lib/with-join-alias "Orders")))]))
+                    (lib/join (lib/join-clause (meta/table-metadata :reviews)
+                                               [(lib/= (meta/field-metadata :reviews :product-id)
+                                                       (meta/field-metadata :orders :product-id))])))
+          ;; Deliberately randomizing the order of the columns.
+          cols  (shuffle (lib/visible-columns query))]
+      (is (=? [{::lib.column-group/group-type :group-type/main}
+               ;; Explicit joins, ordered by join alias.
+               {::lib.column-group/group-type :group-type/join.explicit
+                :join-alias                   "Orders"}
+               {::lib.column-group/group-type :group-type/join.explicit
+                :join-alias                   "Reviews - Product"}
+               ;; Implicit joins, ordered by the join alias and then the FK column name:
+               ;; - "PRODUCT_ID"
+               ;; - "USER_ID"
+               ;; - Orders.PRODUCT_ID and USER_ID
+               ;; - "Reviews - Product.PRODUCT_ID"
+               {::lib.column-group/group-type :group-type/join.implicit
+                :fk-join-alias                nil
+                :fk-field-id                  (meta/id :orders :product-id)}
+               {::lib.column-group/group-type :group-type/join.implicit
+                :fk-join-alias                nil
+                :fk-field-id                  (meta/id :orders :user-id)}
+               {::lib.column-group/group-type :group-type/join.implicit
+                :fk-join-alias                "Orders"
+                :fk-field-id                  (meta/id :orders :product-id)}
```

Comment:
You missed this entry in the comment above

Reply:
Ah, it was included in one line, but that was left over from when the comment was phrased differently and grouped the joins rather than listing each one.

Fixed.